### PR TITLE
enh: validate alignment of sliding windows to fill with taosTimeTruncate and taosTimeCountInterval refactored

### PR DESCRIPTION
**Pull request by bgzhao66**

### CODE REVIEW

File: src/os/src/detail/osTime.c [hunk 1]
```diff
@@ -614,36 +614,25 @@ int64_t taosTimeTruncate(int64_t timeStamp, const SInterval* pInterval, int32_t
 
     if (pInterval->slidingUnit == 'y') {
       tm.tm_mon = 0;
-      tm.tm_year = (int)(tm.tm_year / pInterval->sliding * pInterval->sliding);
+      tm.tm_year -= (int)(tm.tm_year%pInterval->sliding);
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:624]: The effect of the new statement keeps the same as the old code. It is changed to unify the method of truncating aligned local time intervals.

File: src/os/src/detail/osTime.c [hunk 2]
```diff
@@ -614,36 +614,25 @@ int64_t taosTimeTruncate(int64_t timeStamp, const SInterval* pInterval, int32_t
...
     } else {
       int mon = tm.tm_year * 12 + tm.tm_mon;
-      mon = (int)(mon / pInterval->sliding * pInterval->sliding);
+      mon -= (int)(mon % pInterval->sliding);
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:627]: The effect of the new statement keeps the same as the old code. It is changed to unify the method of truncating aligned local time intervals.

File: src/os/src/detail/osTime.c [hunk 3]
```diff
@@ -614,36 +614,25 @@ int64_t taosTimeTruncate(int64_t timeStamp, const SInterval* pInterval, int32_t
...
-      char**  tzname = _tzname;
+    // see https://docs.microsoft.com/en-us/cpp/c-runtime-library/daylight-dstbias-timezone-and-tzname?view=vs-2019
+    int64_t timezone = _timezone;
+    int32_t daylight = _daylight;
+    char**  tzname = _tzname;
   #endif
 
-      start += (int64_t)(timezone * TSDB_TICK_PER_SECOND(precision));
-    }
+    int64_t tz_offset = -1 * timezone * TSDB_TICK_PER_SECOND(precision);
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:634]: Truncate the time per local timezone correctly. For example, Epoch time (ms) 1660608000000 in ASCII string as "Tue Aug 16 08:00:00 2022 CST" should be truncated to 1660579200000 as "Tue Aug 16 00:00:00 2022 CST".

File: src/os/src/detail/osTime.c [hunk 4]
```diff
@@ -614,36 +614,25 @@ int64_t taosTimeTruncate(int64_t timeStamp, const SInterval* pInterval, int32_t
...
+    start -= (start+tz_offset)%pInterval->sliding;
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:639]: Truncate by subtracting the remainder within aligned intervals.

**Review by bgzhao66 — COMMENTED** [review]
post comments to explain code change.

File: src/client/src/tscStream.c [hunk 5]
```diff
@@ -335,6 +335,10 @@ static void tscProcessStreamRetrieveResult(void *param, TAOS_RES *res, int numOf
     tfree(pSql->pSubs);
     pSql->subState.numOfSub = 0;
     pTableMetaInfo->vgroupList = tscVgroupInfoClear(pTableMetaInfo->vgroupList);
+
```
Comment (bgzhao66) [inline src/client/src/tscStream.c:338]: Set the time (stime) of pStream to start if it unset yet.

File: src/query/src/qFill.c [hunk 6]
```diff
@@ -497,13 +507,16 @@ int64_t getNumOfResultsAfterFillGap(SFillInfo* pFillInfo, TSKEY ekey, int32_t ma
   int64_t numOfRes = -1;
   if (numOfRows > 0) {  // still fill gap within current data block, not generating data after the result set.
     TSKEY lastKey = tsList[pFillInfo->numOfRows - 1];
+    if(!validateAlignOfSlidingWindows(lastKey, pFillInfo->currentKey, pFillInfo)) return 0;
```
Comment (bgzhao66) [inline src/query/src/qFill.c:510]: Validate the alignment of sliding windows. If violated, print descriptive messages for trouble shooting along with assert coredump.

File: src/os/src/detail/osTime.c (hunk 4 above)
Comment (bgzhao66) [reply src/os/src/detail/osTime.c:639]: The effect of code keeps exactly the same as old, while readability of the code improves significantly.

**Comment by bgzhao66** [comment]
To improve readability of code and troubleshooting info of assert violation significantly.

File: src/os/src/detail/osTime.c [hunk 7]
```diff
@@ -557,23 +557,29 @@ int64_t taosTimeSub(int64_t timeStamp, int64_t duration, char unit, int32_t prec
   return (int64_t)(mktime(&tm) * TSDB_TICK_PER_SECOND(precision));
 }
 
-int32_t taosTimeCountInterval(int64_t skey, int64_t ekey, int64_t interval, char unit, int32_t precision) {
+int64_t taosTimeTzOffset(int64_t intervalUnit, int32_t precision) {
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:560]: Extract the snippet of code to compute timezone offset in seconds as a standalone function.

File: src/os/src/detail/osTime.c [hunk 8]
```diff
@@ -557,23 +557,29 @@ int64_t taosTimeSub(int64_t timeStamp, int64_t duration, char unit, int32_t prec
...
+    #if defined(WINDOWS) && _MSC_VER >= 1900
+    // see https://docs.microsoft.com/en-us/cpp/c-runtime-library/daylight-dstbias-timezone-and-tzname?view=vs-2019
+    int64_t timezone = _timezone;
+    #endif
+    tz_offset = -1 * timezone * TSDB_TICK_PER_SECOND(precision);
+  }
+  return tz_offset;
+}
+
+int32_t taosTimeCountInterval(int64_t skey, int64_t ekey, int64_t sliding, int64_t slidingUnit, int64_t intervalUnit, int32_t precision) {
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:572]: Refactor taosTimeCountInterval to respect the timezone offset according to intervalUnit, keeping in sync with behaviours of taosTimeTruncate.

**Review by DuanKuanJun — APPROVED** [review]

**Pull request merged**